packages/actor-extract-links-extract-tree/lib/ActorExtractLinksTree.ts

```diff
@@ -42,21 +42,51 @@ export class ActorExtractLinksTree extends ActorExtractLinks {
         this.getTreeQuadsRawRelations(quad,
           currentNodeUrl,
           pageRelationNodes,
-          nodeLinks));
+          nodeLinks,
+          relationDescriptions));
 
       // Resolve to discovered links
       metadata.on('end', () => {
         // Validate if the node forward have the current node as implicit subject
         for (const [ nodeValue, link ] of nodeLinks) {
           if (pageRelationNodes.has(nodeValue)) {
-            links.push({ url: link });
+            const relationDescription = relationDescriptions.get(nodeValue);
+            if (typeof relationDescription !== 'undefined') {
+              relations.push(collectRelation(relationDescription, link));
+            } else {
+              relations.push(collectRelation({}, link));
+            }
           }
         }
-        resolve({ links });
+
+        const node: INode = { relation: relations, subject: currentNodeUrl };
+        let acceptedRelation = relations;
+        this.mediatorOptimizeLinkTraversal.mediate(
+          { treeMetadata: node, context: action.context },
+        )
+          .then(linkTraversalOptimisation => {
+            if (typeof linkTraversalOptimisation !== 'undefined') {
+              acceptedRelation = this.handleOptimization(linkTraversalOptimisation, relations);
+            }
+            resolve({ links: acceptedRelation.map(el => ({ url: el.node })) });
+          })
+          .catch(() => {
+            resolve({ links: acceptedRelation.map(el => ({ url: el.node })) });
+          });
       });
     });
   }
 
+  private handleOptimization(linkTraversalOptimisation: IActorOptimizeLinkTraversalOutput,
+    relations: IRelation[]): IRelation[] {
+    if (typeof linkTraversalOptimisation.filters !== 'undefined') {
```

Comment:
Or perhaps even this to simplify the following.
```suggestion
    if (linkTraversalOptimisation.filters?.size > 0) {
```